### ISSUE TITLE
Cannot configure availabe Webmin modules after account creation using Virtualmin

### ISSUE THREAD

**Issue by shoulders** [virtualmin/virtualmin-gpl#800]
|  SYSTEM INFORMATION||
|----------------------|---------------------------|
|  OS type and version |     Ubuntu Linux 22.04.4     |
|  Webmin version      |     2.111      |
|  Usermin version     |     2.010     |
|  Virtualmin version  |     7.10.0       |
|  Theme version       |     21.10      |
|  Package updates     |      3  package updates are available   |

### Background

I wanted to give one of my user accounts access to the terminal so when I logged into that account I could use the terminal rather than doing everything through my admin account.

The availability of the terminal menu is done in: `Virtualmin --> System Settings --> Server Templates --> Default -> Administrators Webmin modules.`

![dc72b8485fa82e86d750d4b7f389cb963c87066b_2_690x394](https://github.com/virtualmin/virtualmin-gpl/assets/319997/fc2b5239-8f71-49bd-855c-3fd7b829a3c4)

### The issue

There is no option in Virtualmin to enable the terminal after virtual server creation so if you don't set it then you cannot enable it through Virtualmin and without an exhaustive look I guess all of the other `Administrators Webmin modules`

### workaround

For a user to login to Webmin at all they have to be a Webmin user. If they are a Webmin user you can control their modules under

Webmin Users --> Edit Webmin User --> Available Webmin modules

but this present the valid warning message

![image](https://github.com/virtualmin/virtualmin-gpl/assets/319997/3261bc37-ab72-4369-a371-166d23034b9c)

NB: once you have enabled `Terminal` using this method the terminal menu items will remain present.

### Unexpected behaviour

- when I through webmin enabled the `Terminal` the menu items appeared for them user. Once the terminal has been enabled here, the menu items will always be present irrespective of the current user permissions to the terminal.
  ![image](https://github.com/virtualmin/virtualmin-gpl/assets/319997/fd5c59f7-d26c-4a48-816d-a0367fbac092)
- When you enable or disable the `Terminal` in the  `Administrators Webmin modules` section in the relevant server template (i.e. `Developer`, this will enable and disable  (toggle) the setting `Webmin --> Webmin Users --> theuser -->Available Webmin modules --> Tools -->Terminal`
  __  ![6a2bdf23-9376-4c14-bdd4-1d87b364d708](https://github.com/virtualmin/virtualmin-gpl/assets/319997/de6ae2b9-2b99-495e-b76a-627e40405267)

**What does these mean?**
- The Terminal permission is being actively read from the Server Template, in my case the `Developer` template, not the default although it inherits from it.
- when you save changes to the server template this must be pushing changes to users or groups to reflect the changes
- the menu items visibility is set on a one-time basis and then is never checks as to weather it should still show and gives rise to this error message
  [image](https://github.com/virtualmin/virtualmin-gpl/assets/319997/6a2bdf23-9376-4c14-bdd4-1d87b364d708)

**Further Checking**

![image](https://github.com/virtualmin/virtualmin-gpl/assets/319997/83a836af-3045-4ee6-b69a-e7c61bf1233d)

I checked this behaviour with the `BIND DNS Server` and when I enable/disable the option in `Administrators Webmin modules` the permissions are toggle indicating these are actively read possible for all of the setting.

I suspect that menu creation is only done at server creation time or when the feature is enabled via webmin for the first time, but there are no routines to disable the menus and that just enabling the feature in `Administrators Webmin modules` might not create the required menus.

### Proposed solution
- `Administrators Webmin modules` should be moved to the `Account Plan` as it is about permissions and tasks the account owner can do rather than the setup of the server.
- Add some code to disable menus when the user does not have permission.

_This highlights some of the passive and active mix of content in the server template system._

### Additional 
- https://forum.virtualmin.com/t/changing-an-owners-webmin-modules-after-intial-virtual-server-setup/125995
- https://forum.virtualmin.com/t/terminal-works-fine-except-for-one-virtual-server/126252

**Comment by jcameron** [comment]
Hmmm .. what should happen is when the template is updated to change modules, all domains on that template should have their admin user permissions immediately updated to match. This is is unfortunately inconsistent with the way all other template options behave, and confusing! Which suggests that we should fix it...

**Comment by shoulders** [comment]
> what should happen is when the template is updated to change modules, all domains on that template should have their admin user permissions immediately updated to match.

This is happening I think, and as you clarified is expected behaviour, it might just be an issue with the menu items not being hidden when the permissions are removed.

## additional
A note on the relevant server template section saying these are active settings and not fire and forget might be useful.

**Comment by iliajie** [comment]
> This is is unfortunately inconsistent with the way all other template options behave, and confusing! Which suggests that we should fix it.

It would be great to take it out of templates. Where do you think it can go?

**Comment by jcameron** [comment]
Account plan is more like for limiting features for billing purposes though. I think it should remain in the template, but we can add an option elsewhere to change the modules for existing domains..

**Comment by shoulders** [comment]
What about edit owner permission

**Comment by jcameron** [comment]
Yeah that could work

**Comment by shoulders** [comment]
but a quick thought before changing anything. the way it is now all accounts are getting these values on a live basis which means when the template is changed, so is all of the accounts that belong to it, this is a better solution than being set as a one time deal when the virtual server is created and/or having to deal with overriding per account.

So maybe leave it as it is if not being moved to account plan, but just fix the menu display issue.

**Comment by iliajie** [comment]
Using either would offer more a more granular control over the modules available to each domain owner than it is now; and given the intersection of account plans with the capabilities and features configured for owners, the modules available to domain owners (i.e. functionality moved from templates) should be consistent with this pattern.

**Comment by shoulders** [comment]
I have looked at the `account plans` it is loaded with permissions well past what is need for billing.

![image](https://github.com/virtualmin/virtualmin-gpl/assets/319997/d8f8c47a-ccb0-48b5-bb30-452734aad556)

Another section here called `Webmin Modules` seems right to me, but I will leave this choice to you guys.

**Comment by iliajie** [comment]
Yes, as I already mentioned, it will have to intersect with the owner's limits page, and also be configurable in account plans, I think the same way we do for allowed capability section.

Anyway, thanks for bringing this up! We had to address this a long time ago.

**Comment by iliaross** [comment]
Reopened, as this was closed by an automated cleanup run. @shoulders, sorry for the noise, and thanks for the contribution. We definitely didn't mean any disrespect.

If it turns out the issue is already resolved for you, you're welcome to close it.
